torch/csrc/generic/StorageSharing.cpp

```diff
@@ -248,52 +292,153 @@ static PyObject * THPStorage_(shareCuda)(THPStorage *self)
+  PyObject *_ref_counter_offset = PyTuple_GET_ITEM(args, 5);
+  PyObject *_event_handle = PyTuple_GET_ITEM(args, 6);
+  if (!(THPUtils_checkLong(_device) && THPUtils_checkLong(_size_bytes) &&
+        (_handle != Py_None && PyBytes_Check(_handle)) &&
+        (_ref_counter != Py_None && PyBytes_Check(_ref_counter)) &&
+        (_event_handle != Py_None && PyBytes_Check(_event_handle)) &&
+        THPUtils_checkLong(_offset_bytes) &&
+        THPUtils_checkLong(_ref_counter_offset))) {
+    THPUtils_invalidArguments(
+        args,
+        nullptr,
+        "_new_shared in CUDA mode",
+        1,
+        "(int device, bytes handle, int storage_size_bytes, int storage_offset_bytes, bytes _ref_counter, int _ref_counter_offset, bytes event_handle)");
     return nullptr;
   }
 
-  // Storage constructor requires size in numel.
   size_t storage_size = (size_t)THPUtils_unpackLong(_size_bytes) / sizeof(scalar_t);
   ptrdiff_t storage_offset_bytes = (ptrdiff_t)THPUtils_unpackLong(_offset_bytes);
 
   int64_t device = THPUtils_unpackLong(_device);
   at::cuda::CUDAGuard device_guard(device);
 
-  char *buffer;
-  Py_ssize_t handle_size;
-  if (PyBytes_AsStringAndSize(_handle, &buffer, &handle_size) == -1) {
-    return nullptr;
-  }
-  THPUtils_assert(handle_size == CUDA_IPC_HANDLE_SIZE, "incorrect handle size");
-  std::string s_handle = std::string(buffer, handle_size);
+#ifndef __HIP_PLATFORM_HCC__
+  // Ensure that producer prepared all tensor's data
+  std::string s_ipc_event_handle =
+      THPStorage_(bytesAsHandleString)(_event_handle);
+  auto ipc_event_handle =
+      reinterpret_cast<const cudaIpcEventHandle_t*>(s_ipc_event_handle.c_str());
+  cudaEvent_t event;
+  cudaIpcOpenEventHandle(&event, *ipc_event_handle);
+  cudaEventSynchronize(event);
+#else
+  // Already syncronized inside producer stream
+#endif
+
+  std::string s_handle = THPStorage_(bytesAsHandleString)(_handle);
   std::shared_ptr<void> basePtr = c10::cuda::CUDACachingAllocator::getIpcDevPtr(s_handle);
 
   // Offset the basePtr to reconstruct the real storage
   // devPtr = basePtr + storage_offset
   void* devPtr = basePtr.get();
   devPtr = (char*)devPtr + storage_offset_bytes;
 
+  std::string ref_counter_handle = PyBytes_AS_STRING(_ref_counter);
+  ptrdiff_t ref_counter_offset = (ptrdiff_t)THPUtils_unpackLong(_ref_counter_offset);
+
+  auto c = new torch::CudaIPCReceivedData(std::move(basePtr));
+  auto sp = std::shared_ptr<void>(
+      (void*)c, [ref_counter_handle, ref_counter_offset, device](void* ptr) {
+        delete static_cast<torch::CudaIPCReceivedData*>(ptr);
+        // Sync default stream to make sure all operations related to the storage is
+        // finished (otherwise another process may reuse memory and corrupt
+        // data)
+
+        // TODO: Instead of cudaStreamSynchronize it is possibble to add Stream
+        // Callback and release counter inside of it (need to check performance impact)
```

Comment:
I think it would be nice to also include here the discussion we had about whether or not we could have the producer send an un-triggered event to the consumer, and use that to signal if things were done or not. We concluded this was not possible under the current CUDA API, but it's a potential improvement that someone in the future might like to do.